pilot/pkg/model/network.go

```diff
@@ -71,28 +72,67 @@ func NewNetworkManager(env *Environment, xdsUpdater XDSUpdater) (*NetworkManager
 		return nil, err
 	}
 	mgr := &NetworkManager{env: env, NameCache: nameCache, xdsUpdater: xdsUpdater}
-	env.AddNetworksHandler(mgr.reloadAndPush)
-	env.AppendNetworkGatewayHandler(mgr.reloadAndPush)
-	nameCache.AppendNetworkGatewayHandler(mgr.reloadAndPush)
+	env.AddNetworksHandler(mgr.reloadGateways)
+	env.AddNetworksHandler(mgr.reloadNetworkEndpoints)
+	env.AppendNetworkGatewayHandler(mgr.reloadGateways)
+	nameCache.AppendNetworkGatewayHandler(mgr.reloadGateways)
 	mgr.reload()
 	return mgr, nil
 }
 
-func (mgr *NetworkManager) reloadAndPush() {
+// reloadGaeways reloads NetworkGateways and triggers a push if they change.
+func (mgr *NetworkManager) reloadGateways() {
 	mgr.mu.Lock()
-	defer mgr.mu.Unlock()
 	oldGateways := make(NetworkGatewaySet)
 	for _, gateway := range mgr.allGateways() {
 		oldGateways.Add(gateway)
 	}
+
 	changed := !mgr.reload().Equals(oldGateways)
+	mgr.mu.Unlock()
 
 	if changed && mgr.xdsUpdater != nil {
 		log.Infof("gateways changed, triggering push")
 		mgr.xdsUpdater.ConfigUpdate(&PushRequest{Full: true, Reason: []TriggerReason{NetworksTrigger}})
 	}
 }
 
+// reloadNetworkEndpoints reloads NetworkEndpoints and triggers a push if they change.
+func (mgr *NetworkManager) reloadNetworkEndpoints() {
```

Comment:
sorry one more comment:  Missing from the original pr, seems this should be protected with EnableHCMInternalNetworks flag

Reply:
Ah..Good catch